### ISSUE TITLE
whatwg-fetch and breaking boilerplate tests through yarn rw tests

### ISSUE THREAD

**Issue by ogdenstudios** [redwoodjs/redwood#1581]
Hey everyone, I see this has come up a few times before, notably in #1264 

I am still getting these errors even after running `yarn rw upgrade` about five minutes ago. 

`yarn rw test` is failing with this error: 

```
  ● Test suite failed to run

    Cannot find module 'whatwg-fetch' from '../node_modules/@redwoodjs/core/dist/configs/browser/jest.setup.js'

      at Resolver.resolveModule (../node_modules/jest-resolve/build/index.js:306:11)
      at Object.<anonymous> (../node_modules/@redwoodjs/core/dist/configs/browser/jest.setup.js:5:1)
```

Looks like I'm running v0.21.0. All of my web tests are failing, including the boilerplate ones from running the page generator like: `yarn rw g page home`. API tests are all fine. 

Are there additional steps I need to take in upgrading, beyond `yarn rw upgrade`? Or is something still broken here? Thank you!

**Comment by Tobbe** [comment]
Hi @ogdenstudios, sorry you're having problems running the tests.

I just did a quick *ehrm* test... on my side, and the test of the tests seem to work just fine

![image](https://user-images.githubusercontent.com/30793/102026725-ce968f00-3d9f-11eb-953e-dde681b0fa40.png)

I took a quick glance at the issue you linked, but didn't see anything about whatwg-fetch. What did I miss?

Just for completeness sake, can you please post the output of `yarn rw info`?

**Comment by ogdenstudios** [comment]
Ope! Sorry. That issue I linked was linked from #985. It looks like 985 was closed in favor of 1264. That's why I linked there, but I found 1264 after searching for `whatwg-fetch` in the issue tracker. 

Here's my `yarn rw info` output: 

```
  System:
    OS: macOS 11.0.1
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 15.2.1 - /var/folders/cp/d4ldtj4s6vn7nw41txkf9j640000gn/T/yarn--1607906589711-0.9236027698502045/node
    Yarn: 1.22.10 - /var/folders/cp/d4ldtj4s6vn7nw41txkf9j640000gn/T/yarn--1607906589711-0.9236027698502045/yarn
  Databases:
    SQLite: 3.32.3 - /usr/bin/sqlite3
  Browsers:
    Chrome: 87.0.4280.88
    Edge: 87.0.664.60
    Firefox: 83.0
    Safari: 14.0.1
  npmPackages:
    @redwoodjs/core: ^0.21.0 => 0.21.0 

✨  Done in 1.75s.
```

**Comment by Tobbe** [comment]
So both you and the guy in #985 are on macOS, but so is a lot of other people... But that's about the only common denominator I can see from the info output.

What do you get if you run `yarn why whatwg-fetch`?
This is my output
```
$ yarn why whatwg-fetch
yarn why v1.22.4
[1/4] Why do we have the module "whatwg-fetch"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "whatwg-fetch@3.5.0"
info Reasons this module exists
   - "_project_#api#@redwoodjs#api#graphql-tools#@graphql-tools#relay-operation-optimizer#relay-compiler#fbjs#isomorphic-fetch" depends on it
   - Hoisted from "_project_#api#@redwoodjs#api#graphql-tools#@graphql-tools#relay-operation-optimizer#relay-compiler#fbjs#isomorphic-fetch#whatwg-fetch"
info Disk size without dependencies: "68KB"
info Disk size with unique dependencies: "68KB"
info Disk size with transitive dependencies: "68KB"
info Number of shared dependencies: 0
Done in 2.04s.
```

**Comment by ogdenstudios** [comment]
Well this is probably the root of the issue: 

```
yarn why whatwg-fetch
yarn why v1.22.10
[1/4] 🤔  Why do we have the module "whatwg-fetch"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
error We couldn't find a match!
✨  Done in 1.20s.
```

I double checked in `yarn.lock` and found nothing. Removed all my `node_modules` folders and ran `yarn install` fresh, had the same response with `yarn why whatwg-fetch`.

**Comment by peterp** [comment]
@ogdenstudios Could you remove `yarn.lock` and `node_modules` and try again? I think the way we run the upgrade command sometimes introduces a broken lock file.

**Comment by peterp** [comment]
In order to fix the upgrade command I don't think we should rely on yarn's upgrade command, but rather:

1. find all the package.json files.
2. find and bump the versions for the `@redwood/` packages.
3. run `yarn install`

**Comment by Tobbe** [comment]
This is from a freshly initiated redwood project

```
tobbe@XPS9550 MINGW64 ~/dev/redwood/whatwgfetch
$ yarn why whatwg-fetch
yarn why v1.22.4
[1/4] Why do we have the module "whatwg-fetch"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "whatwg-fetch@3.5.0"
info Reasons this module exists
   - "_project_#api#@redwoodjs#api#graphql-tools#@graphql-tools#relay-operation-optimizer#relay-compiler#fbjs#isomorphic-fetch" depends on it
   - Hoisted from "_project_#api#@redwoodjs#api#graphql-tools#@graphql-tools#relay-operation-optimizer#relay-compiler#fbjs#isomorphic-fetch#whatwg-fetch"
info Disk size without dependencies: "68KB"
info Disk size with unique dependencies: "68KB"
info Disk size with transitive dependencies: "68KB"
info Number of shared dependencies: 0
Done in 1.68s.
```

I'm not 100% sure how to read the output, but if I understand it correctly yarn believes it has whatwg-fetch installed because isomorphic-fetch depends on it.
It doesn't seem to know that redwood also depends on it.

Is it because redwood only lists whatwg-fetch as a devDependency? Should we move it to a regular dependency?

**Comment by ogdenstudios** [comment]
> @ogdenstudios Could you remove `yarn.lock` and `node_modules` and try again? I think the way we run the upgrade command sometimes introduces a broken lock file.

I removed `yarn.lock` at the root and all my `node_modules` folders. Then I ran `yarn install`. Similar problem here: 

```
yarn why whatwg-fetch
yarn why v1.22.10
[1/4] 🤔  Why do we have the module "whatwg-fetch"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
error We couldn't find a match!
✨  Done in 1.05s.
```

Is it more than the lockfile? The only thing I've done to the `package.json` since setup is run `yarn upgrade` and run the RedwoodJS Tailwind install command. 

[Here is the repo if you want to look more closely](https://github.com/ogdenstudios/final-project-msse-661). I'm in a software engineering master's program and this is for my web dev course - feel free to do with it as you wish, it's just kind of a toy project for class + to try out Redwood. Y'all are doing great things!

**Comment by forresthayes** [comment]
I'm seeing the same error in a new Redwood app with only a generated home page.

```
yarn rw info
  System:
    OS: macOS 11.1
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.11.0 - /var/folders/sy/tjb1z4757xbfrhflgnh9pxsc0000gn/T/yarn--1610406674838-0.6699669243505681/node
    Yarn: 1.22.5 - /var/folders/sy/tjb1z4757xbfrhflgnh9pxsc0000gn/T/yarn--1610406674838-0.6699669243505681/yarn
  Databases:
    SQLite: 3.32.3 - /usr/bin/sqlite3
  Browsers:
    Chrome: 87.0.4280.141
    Edge: 87.0.664.75
    Firefox: 84.0.1
    Safari: 14.0.2
  npmPackages:
    @redwoodjs/core: ^0.22.1 => 0.22.1 
```

**Comment by thedavidprice** [comment]
Wow, gents. This one is bonkers. I cloned the repo and took a look (thanks for providing @ogdenstudios). The package isn't there. Removing `node_modules` and recreating `yarn.lock` does **not** resolve. I also upgraded to v0.21.1, repeated the process, and still. no. `whatwg-fetch` package.

**Here's where it got even more fun.** --> I have an existing project that 1) runs test successfully and 2) has `whatwg-fetch` in the lockfile. (Running v0.21.1)
- I _deleted_ `node_modules` and `yarn.lock` and then ran `yarn`
- ERROR: no `whatwg-fetch` in my new `yarn.lock`

@peterp I'm I correct in thinking there's something missing from our v0.21.1 packages? Must be, right? We've included it:
1. here in the [root package.json](https://github.com/redwoodjs/redwood/blob/9daf39aae11762db39abfbf41083f39be254ef0d/package.json#L49)
2. and [here in core/package.json](https://github.com/redwoodjs/redwood/blob/0288d7a75513e0d95269f46130d212ad49c57f7b/packages/core/package.json#L71)

So very confused...

## Workaround
@ogdenstudios to complete your project you can manually add this package to your project's dev dependencies.
- `yarn add -WD whatwg-fetch`
- `rw test` should then work. Some tests are failing but everything is running correctly.

**Comment by Tobbe** [comment]
As I noted in a previous comment we had whatwg-fetch because one of our dependencies brought it in.

```
=> Found "whatwg-fetch@3.5.0"
info Reasons this module exists
   - "_project_#api#@redwoodjs#api#graphql-tools#@graphql-tools#relay-operation-optimizer#relay-compiler#fbjs#isomorphic-fetch" depends on it
   - Hoisted from "_project_#api#@redwoodjs#api#graphql-tools#@graphql-tools#relay-operation-optimizer#relay-compiler#fbjs#isomorphic-fetch#whatwg-fetch"
```

If you look in that chain of packages you'll find [fbjs](https://github.com/facebook/fbjs). 

In April of last year they replaced isomorphic-fetch with cross-fetch, which doesn't have a dependency on whatwg-fetch

https://github.com/facebook/fbjs/commit/16620e3623fd46b435313abd639b86313facb7b8#diff-7e0047660b998823aee877b216662b68226c4d851056cc51f115fc3a1b6589d9

This was released as fbjs 2.0.0 (up from 1.0.0) in July 2020. And then fbjs 3.0.0 was released in October.

Redwood's dependencies have now updated to use fbjs 3, and so we no longer get whatwg-fetch. This is the reason this doesn't work at all anymore.
